WordPress/src/main/java/org/wordpress/android/ui/uploads/PostUploadNotifier.java

```diff
@@ -33,62 +36,161 @@
+        sNotificationData.totalMediaItems++;
+        startOrUpdateForegroundNotification(null);
     }
 
     void updateNotificationIcon(PostModel post, Bitmap icon) {
-        NotificationData notificationData = sPostIdToNotificationData.get(post.getId());
+        // TODO MEDIA reimplement or remove completely
+
+//        NotificationData notificationData = sPostIdToNotificationData.get(post.getId());
+//
+//        if (icon != null) {
+//            notificationData.latestIcon = icon;
+//            mNotificationBuilder.setLargeIcon(notificationData.latestIcon);
+//        }
+//        doNotify(sPostIdToNotificationData.get(post.getId()).notificationId, mNotificationBuilder.build());
+    }
+
+    void incrementUploadedPostCountFromForegroundNotification(PostModel post) {
+        sNotificationData.currentPostItem++;
+        if (post.isPage()) {
+            sNotificationData.totalPageItemsIncludedInPostCount--;
+        }
 
-        if (icon != null) {
-            notificationData.latestIcon = icon;
-            mNotificationBuilder.setLargeIcon(notificationData.latestIcon);
+        // update Notification now
+        if (!removeNotificationAndStopForegroundServiceIfNoItemsInQueue()) {
+            updateForegroundNotification(post);
         }
-        doNotify(sPostIdToNotificationData.get(post.getId()).notificationId, mNotificationBuilder.build());
     }
 
-    void cancelNotification(PostModel post) {
-        NotificationData notificationData = sPostIdToNotificationData.get(post.getId());
-        if (notificationData != null) {
-            mNotificationManager.cancel(notificationData.notificationId);
+    void incrementUploadedMediaCountFromProgressNotificationOrFinish() {
+        sNotificationData.currentMediaItem++;
+
+        if (!removeNotificationAndStopForegroundServiceIfNoItemsInQueue()) {
+            // update Notification now
+            updateForegroundNotification(null);
         }
-        mService.stopForeground(true);
+    }
+
+    boolean removeNotificationAndStopForegroundServiceIfNoItemsInQueue() {
+        if (sNotificationData.currentPostItem == sNotificationData.totalPostItems
+                && sNotificationData.currentMediaItem == sNotificationData.totalMediaItems) {
+            mNotificationManager.cancel(sNotificationData.notificationId);
+            // reset the notification id so a new one is generated next time the service is started
+            sNotificationData.notificationId = 0;
+            resetNotificationCounters();
+            mService.stopForeground(true);
+            return true;
+        }
+        return false;
+    }
+
+    void resetNotificationCounters() {
+        sNotificationData.currentPostItem = 0;
+        sNotificationData.currentMediaItem = 0;
+        sNotificationData.totalMediaItems = 0;
+        sNotificationData.totalPostItems = 0;
+        sNotificationData.totalPageItemsIncludedInPostCount = 0;
     }
 
     void cancelErrorNotification(PostModel post) {
```

Comment:
There's some inconsistency with the use of `@NonNull` and `@Nullable` in this class - some methods use these annotations, while some (like this one) don't. Do you think this should be addressed?

Reply:
good catch, doing a pass on the public methods in 7b5eabc.
Also made sure these local methods were set to `private` in d90dccd: `removeNotificationAndStopForegroundServiceIfNoItemsInQueue` and `resetNotificationCounters`